README.md
<!--
 * @Author: 李文超
 * @Date: 2021-03-24 17:57:44
 * @LastEditors: 李文超
 * @LastEditTime: 2021-03-25 09:55:16
 * @Description: file content
 * 
 * @FilePath: \auto-html2css\README.md
-->
# auto-html2css README

## 概述

提高工作效率，减少重复劳动用

[代码地址](https://github.com/Yukilwc/auto-html2css)

## 功能

1 自动转换html结构中的class，生成嵌套层级的scss代码

## 使用

> 首先，需要在编辑器中，选中一段代码

1. 命令，ctrl+shift+p,输入指令html2css,并执行
2. 快捷键,ctrl+alt+z,执行

当提示转换完成后，转换后的内容将会在剪切板上

## 配置

> open workspace settings - extensions - AutoHtml2Css ,编辑json文件

| 字段名         | 含义              | 可选值                                                     |
| -------------- | ----------------- | ---------------------------------------------------------- |
| extraClassList | 要排除的class名单 | 字符串类型，使用逗号隔开，会进行切分并转换成正则表达式匹配 |

CHANGELOG.md
# Change Log

All notable changes to the "auto-html2css" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

lib/generate.js
/*
 * @Author: 李文超
 * @Date: 2021-03-22 11:19:24
 * @LastEditors: 李文超
 * @LastEditTime: 2021-03-25 10:05:32
 * @Description: file content
 * @FilePath: \auto-html2css\lib\generate.js
 */
console.log('==========generate',)
const fs = require('fs')
// const data = fs.readFileSync('./template.html', "utf-8")
const cheerio = require('cheerio')
const parseCSS = require('./parseCss')
const getExtraClassList = require('./extraClassList')
// 是否在特殊名单内
const isInExtraClassList = (item, extraClassList) => {
    let res = false
    res = extraClassList.find((extraItem => {
        let isString = Object.prototype.toString.call(extraItem) == "[object String]"
        let isReg = Object.prototype.toString.call(extraItem) == "[object RegExp]"
        if (isString) {
            return extraItem === item
        }
        else if (isReg) {
            return extraItem.test(item)
        }
        else {
            throw new Error('名单只支持字符串和正则表达式')
        }

    }))

    return res
}
const getNodeClass = (node) => {
    if (node.attribs && node.attribs.class) {
        let classList = node.attribs.class.split(/\s+/)
        let filterList = classList.filter((item) => {
            if (!item) return false
            // 去除小程序中class变量,以及名单中的class
            if (/{{.*}}/.test(item)) {
                return false
            }
            // 名单内排除
            else if (isInExtraClassList(item, getExtraClassList())) {
                return false
            }
            else {
                return true
            }
        })
        if (filterList.length > 0) {
            let str = '.' + filterList.join('.')
            return str
        }
        else {
            return ''
        }
    }
    else {
        return ''
    }
}
const getParentLink = (node, link = []) => {
    let nodeClass = getNodeClass(node)
    link.push(nodeClass)
    if (node.parent) {
        getParentLink(node.parent, link)
        return link
    }
    else {
        return link
    }
}

// const classAst = htmlNodeList.map(() => {
//     // 如何做转化？
// })

// fs.writeFileSync('./res.scss', scssStr)
const generate = (htmlStr) => {
    const $ = cheerio.load(htmlStr)
    let scssStr = ''
    const htmlNodeList = $('*').toArray()
    htmlNodeList.forEach((node) => {
        let link = getParentLink(node, []).filter(item => item).reverse()
        if (link.length > 0) {
            let linkStr = '\n' + link.join(' ') + '{\n}\n'
            scssStr += linkStr
        }

    })
    scssStr = parseCSS(scssStr)
    return scssStr
}
module.exports = generate


lib/parseCss.js
const R = require('ramda')
function log(s) {
    if (console) console.log(s);
}
function tr(s) {
    return R.trim(s.replace(/\t+/, ' '));
}

var indentS;
var openingBracket;
var closingBracket;
var semiColumn;
var eol;
var keepOutsideComments;
var removeHacks;
var vars = {};
var comments = {};
var dataURLs = {};

function trimColor(str) {
    return str.replace(/#([0-9a-f]{3}|[0-9a-f]{6})\b/ig, function (a) {
        return a.toLowerCase()
    })
}
// 变量，暂不使用
// function parseVars() {
//     vars = {};
//     variables = $('#variables').val().split(/[\n;]/);
//     $(variables).each(function () {
//         var pair = this.replace(/\/\*[\s\S]*?\*\//gm, '')
//             .replace(/\/\/[\s\S]*$/gm, '');

//         pair = pair.split(/[:=]/);
//         if (pair.length > 1) {
//             vars[trimColor(R.trim(pair[1]))] = R.trim(pair[0]);
//         }
//     })
// }

function initOptions() {
    console.log('==========配置初始化',)
    // 此处设置语言
    var langVal = 'sass';
    if (langVal == 'less') {
        eol = ';';
        openingBracket = '{';
        closingBracket = '}';
        semiColumn = ':';
    } else if (langVal == 'sass') {
        eol = ';';
        openingBracket = '{';
        closingBracket = '}';
        semiColumn = ':';
    } else if (langVal == 'stylus') {
        eol = '';
        openingBracket = '';
        closingBracket = '';
        semiColumn = '';
    }
    // parseVars();
    // 此处设置缩进
    var indentVal = 'indent2';
    if (indentVal == 'indentt') indentS = '\t';
    else if (indentVal == 'indent2') indentS = '  ';
    else if (indentVal == 'indent4') indentS = '    ';
    // 命名空间外的注释
    keepOutsideComments = true;
    // 命名空间内的注释
    keepInsideComments = false;
    // 去除Hacks和前缀
    removeHacks = false;
}

function makeMark(num) {
    var str = "";
    str += '\t\t\t\t\t\t\t\t\t\t\t';
    while (num--) {
        str += ' '
    }
    str += '\t\t\t\t\t\t\t\t\t\t\t';
    return str
}

function decodeMark(str) {
    var num = 0;
    str.replace(/ /g, function () {
        num++
    })
    return num
}

function parseCSS(s) {
    initOptions()
    var least = { children: {} };


    comments = {};
    var i = 1;

    //replace comments with marks;
    s = s.replace(/\/\*[\s\S]*?\*\//gm, function (a) {
        comments[++i] = a;
        return makeMark(i);
    });

    //inside comments to fake declarations;
    s = s.replace(/([^{]+)\{([^}]+)\}/g, function (group, selector, declarations) {

        return selector + '{' + declarations.replace(/\t{10} +\t{10}/g, function (a) {
            a = decodeMark(a);
            return 'comment__-' + a + ': ' + a + ';'
        }) + '}';
    });


    //outside comments to fake selectors;
    s = s.replace(/\t{10} +\t{10}/g, function (a) {
        a = decodeMark(a);
        return '.__comment__-' + a + ' { index: ' + a + ';}'
    });


    dataURLs = {};

    //dataURL to fake url
    s = s.replace(/url\((data:[^\)]+)\)/gm, function (a, dataURL) {
        dataURLs[++i] = dataURL;
        return 'url(__data__' + i + ')';
    })


    s.replace(/([^{]+)\{([^}]+)\}/g, function (group, selector, declarations) {
        var o = {};
        o.source = group;
        o.selector = tr(selector);

        var path = least;

        if (o.selector.indexOf(',') > -1) {
            // Comma: grouped selector, we skip
            var sel = o.selector;
            if (!path.children[sel]) {
                path.children[sel] = { children: {}, declarations: [] };
            }
            path = path.children[sel];
        } else {
            // No comma: we process

            // Fix to prevent special chars to break into parts
            o.selector = o.selector.replace(/\s*([>\+~])\s*/g, ' &$1');
            o.selector = o.selector.replace(/(\w)([:\.])/g, '$1 &$2');

            o.selectorParts = o.selector.split(/[\s]+/);
            for (var i = 0; i < o.selectorParts.length; i++) {
                var sel = o.selectorParts[i];
                // We glue the special chars fix
                sel = sel.replace(/&(.)/g, '& $1 ');
                sel = sel.replace(/& ([:\.]) /g, '&$1');

                if (!path.children[sel]) {
                    path.children[sel] = { children: {}, declarations: [] };
                }
                path = path.children[sel];
            }
        }


        declarations.replace(/([^:;]+):([^;]+)/g, function (decl, prop, val) {

            //remove hacks
            if (removeHacks) {
                decl = decl
                    .replace(/\s*[\*_].*/g, '')
                    .replace(/.*(-webkit-|-o-|-moz-|-ms-|-khtml-|DXImageTransform|\\9|\\0|expression|opacity\s*=).*/g, '')
                if (!R.trim(decl)) {
                    return
                }
            };

            val = trimColor(val);
            if (vars[val]) {
                val = vars[val];
            } else {
                var a = [];
                R.forEach(function () {
                    a.push(vars[this] || this)
                }, val.split(' '));
                val = a.join(' ');
            }

            var declaration = {
                source: decl,
                property: tr(prop),
                value: tr(val)
            };
            path.declarations.push(declaration);
        });
    });

    return exportObject(least);
}
var depth = 0;
var s = '';
function exportObject(path) {
    // console.log('==========path', path)
    var s = '';
    R.forEachObjIndexed(function (val, key) {
        // console.log('==========key val', key, val)
        if (key.slice(0, 12) == '.__comment__') {
            keepOutsideComments && (s += getIndent() + comments[val.declarations[0].value] + '\n');
            return
        } else {
            s += getIndent() + key + ' ' + openingBracket + '\n';
        }
        depth++;
        for (var i = 0; i < val.declarations.length; i++) {
            var decl = val.declarations[i];
            if (decl.property.slice(0, 9) == 'comment__') {
                keepInsideComments && (s += getIndent() + comments[decl.value] + '\n');
            } else {
                s += getIndent() + decl.property + semiColumn + ' ' + decl.value + eol + '\n';
            }

        }
        s += exportObject(val);
        depth--;
        s += getIndent() + closingBracket + '\n';
    }, path.children)
    // Remove blank lines - http://stackoverflow.com/a/4123442
    s = s.replace(/^\s*$[\n\r]{1,}/gm, '');


    s = s.replace(/url\(__data__(\d+)\)/gm, function (a, i) {
        return 'url(' + dataURLs[i] + ')';
    })

    return s;
}
function getIndent() {
    var s = '';
    for (var i = 0; i < depth; i++) {
        s += indentS;
    }
    return s;
}

module.exports = parseCSS

extension.js
/*
 * @Author: 李文超
 * @Date: 2021-03-24 17:57:44
 * @LastEditors: 李文超
 * @LastEditTime: 2021-03-25 09:47:33
 * @Description: file content
 * @FilePath: \auto-html2css\extension.js
 */
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
console.log('==========start',)
const vscode = require('vscode');

const generate = require('./lib/generate')
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "auto-html2css" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with  registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('auto-html2css.html2css', function () {
		// The code you place here will be executed every time your command is executed

		console.log('==========test 1',)
		// console.log('==========configuration ',configuration.get() )
		try {
			handleSelection()
		}
		catch (e) {
			console.error(e.message)
			vscode.window.showErrorMessage(e.message || '转换发生错误')
		}
	});

	context.subscriptions.push(disposable);
}

// this method is called when your extension is deactivated
function deactivate() { }
const handleSelection = () => {
	const editor = vscode.window.activeTextEditor
	const selection = editor.selection
	let text = editor.document.getText(selection)
	let res = generate(text)
	const clipboard = vscode.env.clipboard
	clipboard.writeText(res)
	console.log('==========当前激活的editor',)
	vscode.window.showInformationMessage('转换成功');
}

module.exports = {
	activate,
	deactivate
}
